src/v/storage/disk_log_impl.cc

```diff
@@ -557,46 +558,56 @@ model::term_id disk_log_impl::term() const {
+    offset_stats ret;
+    ret.start_offset = start_offset;
+    if (ret.start_offset > model::offset(0)) {
+        ret.dirty_offset = ret.start_offset - model::offset(1);
+        ret.committed_offset = ret.dirty_offset;
+    }
+    return ret;
+}
 
 offset_stats disk_log_impl::offsets() const {
-    if (_segs.empty()) {
-        offset_stats ret;
-        ret.start_offset = _start_offset;
-        if (ret.start_offset > model::offset(0)) {
-            ret.dirty_offset = ret.start_offset - model::offset(1);
-            ret.committed_offset = ret.dirty_offset;
-        }
-        return ret;
-    }
-    // NOTE: we have to do this because ss::circular_buffer<> does not provide
-    // with reverse iterators, so we manually find the iterator
-    segment_set::type end;
-    segment_set::type term_start;
-    for (int i = (int)_segs.size() - 1; i >= 0; --i) {
-        auto& seg = _segs[i];
-        if (!seg->empty()) {
-            if (!end) {
-                end = seg;
-            }
-            // find term start offset
-            if (seg->offsets().term < end->offsets().term) {
-                break;
-            }
-            term_start = seg;
-        }
-    }
-    if (!end) {
-        offset_stats ret;
-        ret.start_offset = _start_offset;
-        if (ret.start_offset > model::offset(0)) {
-            ret.dirty_offset = ret.start_offset - model::offset(1);
-            ret.committed_offset = ret.dirty_offset;
-        }
-        return ret;
+    /**
+     * We have to iterate backward over segments set as we have to find last
+     * non empty segment, last segment with initialized committed offset and
+     * first offset of current term
+     */
+    const auto rend = std::make_reverse_iterator(_segs.cbegin());
+    const auto rbegin = std::find_if(
+      std::make_reverse_iterator(_segs.cend()),
+      rend,
+      [](const segment_set::type& seg) { return !seg->empty(); });
+
+    // If all segments are empty, return empty offsets.
+    if (rend == rbegin) {
+        return empty_offsets(_start_offset);
     }
+
+    auto it_last_committed = std::find_if(
+      rbegin, rend, [](const segment_set::type& seg) {
+          return seg->offsets().committed_offset >= model::offset(0);
+      });
+    // If nothing was committed, use the first segment
+    const auto& last_committed = it_last_committed != rend ? *it_last_committed
+                                                           : _segs.front();
+
+    // find last segment with term that is smaller than current one
+    auto it_term_start = std::lower_bound(
+      _segs.begin(),
+      _segs.end(),
+      (*rbegin)->offsets().term,
+      [](const segment_set::type& seg, model::term_id term) {
+          return seg->offsets().term < term;
+      });
+    // If term_start not found, use the first segment
+    const auto& term_start = it_term_start != _segs.end() ? *it_term_start
+                                                          : _segs.front();
```

Comment:
it seems like the code above that was deleted during this clean-up would always select term_start to initially be the first non-empty segment found in reverse iteration order. that doesn't seem to be the preserved semantics here, but its a bit hard to map the semantics exactly between the versions of this code.

Reply:
The old code would iterate in reverse, updating `term_start` at the end of the loop, until it finds a term less than the search value, at which point it would break from the loop, skipping the update. It's the same semantics as `lower_bound`, but linear rather than binary search. I posited that binary search would be faster, in particular because the comparison is at the end of a pointer.  We (mostly @mmaslankaprv) [wrote a quick benchmark](https://quick-bench.com/q/GM9TTPB0tfHsSym4y2hyNR2Waqs).

Reply:
I see now how they are equivalent. Same question about `segs.front` as above